src/connector/src/source/filesystem/opendal_source/opendal_reader.rs

```diff
@@ -0,0 +1,172 @@
+            let source_ctx = self.source_ctx.clone();
+
+            let split_id = split.id();
+
+            let data_stream =
+                Self::stream_read_object(self.connector.op.clone(), split, self.source_ctx.clone());
+
+            let parser =
+                ByteStreamSourceParserImpl::create(self.parser_config.clone(), source_ctx).await?;
+            let msg_stream = if matches!(
+                parser,
+                ByteStreamSourceParserImpl::Json(_) | ByteStreamSourceParserImpl::Csv(_)
+            ) {
+                parser.into_stream(nd_streaming::split_stream(data_stream))
+            } else {
+                parser.into_stream(data_stream)
+            };
+            #[for_await]
+            for msg in msg_stream {
+                let msg = msg?;
+                self.source_ctx
+                    .metrics
+                    .partition_input_count
+                    .with_label_values(&[&actor_id, &source_id, &split_id])
+                    .inc_by(msg.chunk.cardinality() as u64);
+                yield msg;
+            }
+        }
+    }
+
+    #[try_stream(boxed, ok = Vec<SourceMessage>, error = anyhow::Error)]
+    pub async fn stream_read_object(
+        op: Operator,
+        split: OpendalFsSplit<C>,
+        source_ctx: SourceContextRef,
+    ) {
+        let actor_id = source_ctx.source_info.actor_id.to_string();
+        let source_id = source_ctx.source_info.source_id.to_string();
+        let max_chunk_size = source_ctx.source_ctrl_opts.chunk_size;
+        let split_id = split.id();
+
+        let object_name = split.name.clone();
+
+        let reader = op
+            .reader_with(&object_name)
+            .range(split.offset as u64..)
+            .await?;
+
+        let stream_reader = StreamReader::new(
+            reader.map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e)),
+        );
+        let buf_reader = BufReader::new(stream_reader);
+        let stream = ReaderStream::with_capacity(buf_reader, STREAM_READER_CAPACITY);
+
+        let mut offset: usize = split.offset;
+        let mut batch_size: usize = 0;
+        let mut batch = Vec::new();
+        #[for_await]
+        for read in stream {
+            let bytes = read?;
+            let len = bytes.len();
+            let msg = SourceMessage {
+                key: None,
+                payload: Some(bytes.as_ref().to_vec()),
+                offset: offset.to_string(),
+                split_id: split.id(),
+                meta: SourceMeta::Empty,
+            };
+            offset += len;
+            batch_size += len;
+            batch.push(msg);
+            if batch.len() >= max_chunk_size {
+                source_ctx
+                    .metrics
+                    .partition_input_bytes
+                    .with_label_values(&[&actor_id, &source_id, &split_id])
+                    .inc_by(batch_size as u64);
+                batch_size = 0;
+                yield batch.clone();
```

Comment:
clone maybe too expensive here?

Reply:
Just keep the same as the previous implementation(https://github.com/risingwavelabs/risingwave/blob/5ad55af49305a6e9f90d4ce30da87b72a841f98b/src/connector/src/source/filesystem/s3/source/reader.rs#L128), let me see whether it can be removed.